hydra_agent/graphutils_operations.py

```diff
@@ -0,0 +1,147 @@
+import urllib.request
+import json
+import logging
+from urllib.error import URLError, HTTPError
+from redis_proxy import RedisProxy
+from graphutils import GraphUtils
+from redisgraph import Graph, Node
+from requests import Session
+
+logger = logging.getLogger(__file__)
+
+
+class GraphOperations(Session):
+
+    def __init__(self, entrypoint_url, redis_connection):
+        self.entrypoint_url = entrypoint_url
+        self.redis_connection = redis_connection
+        self.connection = redis_connection.get_connection()
+        self.vocabulary = 'vocab'
+        self.graph_utils = GraphUtils(redis_connection)
+        self.redis_graph = Graph("apidoc", redis_connection)
+        super().__init__()
+
+    def get_processing(self, url: str, resource: dict) -> None:
+        """Synchronize Redis upon new GET operations
+        :param url: Resource URL to be updated in Redis.
+        :param resource: Resource object fetched from server.
+        :return: None.
+        """
+        # Receiving updated object from the Server
+        json_response = super().get(url).json()
+
+        url = url.rstrip('/').replace(self.entrypoint_url, "EntryPoint")
+        # Updating Redis
+        # First case - When processing a GET for a resource
+        try:
+            entrypoint, resource_endpoint, resource_id = url.split('/')
+
+            # Building the the collection id, i.e. vocab:Entrypoint/Collection
+            redis_collection_id = self.vocabulary + \
+                ":" + entrypoint + \
+                "/" + resource_endpoint
+
+            collection_members = self.graph_utils.read(
+                match="collection",
+                where="id='{}'".format(redis_collection_id),
+                ret="members")
+
+            # Accessing the members with redis-set response structure
+            collection_members = eval(collection_members[0][1][0].decode())
+            collection_members.append({'@id': resource['@id'],
+                                       '@type': resource['@type']})
+
+            # Updating the collection properties with the nem member
+            self.graph_utils.update(
+                match="collection",
+                where="id='{}'".format(redis_collection_id),
+                set="members = \"{}\"".format(str(collection_members)))
+            # Creating node for new collection member
+            node = self.graph_utils.add_node("objects" + resource['@type'],
```

Comment:
Spent some time into this but for some reason couldn't get it to set properly the relation and properties in the graph yet. It just doesn't add the properties correctly. Will comment on the solution here soon.

Reply:
https://github.com/HTTP-APIs/hydra-python-agent/issues/110

Reply:
To properly create valid edges(relations) between existing nodes in RedisGraph(and not create nodes again) one should:
```
MATCH (f:%s{%s:'%s'}), (t:%s{%s:'%s'}) CREATE (f)-[:in]->(t)
```
With DroneCollection example:
```
GRAPH.QUERY apidoc "MATCH (s:collection {type:'DroneCollection'} ), (d:objectsDrone {id:'/serverapi/DroneCollection/ea7e438e-a93d-436d-a7e9-994c13d49dc0'} ) CREATE (s)-[:has_Drone]->(d)"
```
This is not available yet on the oficial doc and [one issue is open about it in RedisGraph repo](https://github.com/RedisGraph/redisgraph-py/issues/16). Will add to our internal RedisGraph doc that I'm populating.

Reply:
the issue states that `MATCH (f:%s{%s:'%s'}), (t:%s{%s:'%s'}) CREATE (f)-[:in]->(t)` is working though with some memory cleanup problem.

Reply:
Yeah.. I've been finding RedisGraph really awesome, think it fits really well with the Agent and will enable powerful things in the future. However, like Hydra, it's not really a finished product and has some limitations and enhancements on the way. Linking existing two nodes is somewhat a basic feature but still has these bits of loose ends/not so straight forward as desired.

Also, checking the graph for modifications during implementation takes a lot of query writing and time. They're still working on a visualization tool as mentioned in this issue https://github.com/RedisGraph/RedisGraph/issues/35 which, depending on the way it gets implemented, might be a stepping stone for a more visual Demo for the Agent. And ofc saving time when developing changes to the Agent and checking what's actually happening.